workflows-cli/src/command_runner.rs

```diff
@@ -0,0 +1,143 @@
+    /// Returns a CommandLike runner
+    fn new_command(&self, cmd: &str) -> Box<dyn CommandLike>;
+}
+
+/// Return a real command runner
+pub struct RealCommandFactory;
+
+/// Heap allocated command to assign command runner at run-time
+impl CommandFactory for RealCommandFactory {
+    fn new_command(&self, cmd: &str) -> Box<dyn CommandLike> {
+        Box::new(RealCommand {
+            cmd: Command::new(cmd),
+        })
+    }
+}
+
+/// Structure of mock entry as in mapping file
+#[derive(Debug, Deserialize, Default, Clone)]
+struct MockEntry {
+    /// Calling command
+    command: String,
+    /// CLI return
+    response: String,
+    /// Exit code
+    code: i32,
+}
+
+/// Required fields to mock real command
+pub struct MockCommand {
+    /// Command being used
+    cmd: String,
+    /// CLI arguents
+    args: Vec<String>,
+    /// Map CLI call to output
+    mappings: Vec<MockEntry>,
+}
+
+impl MockCommand {
+    /// Initiate mock command runner
+    pub fn new(cmd: &str) -> Self {
+        let yaml_str = fs::read_to_string("./tests/mock_commands.yaml")
+            .expect("Failed to read mock_commands.yaml");
+
+        let mappings: HashMap<String, Vec<MockEntry>> =
+            serde_yaml::from_str(&yaml_str).expect("Failed to parse YAML");
+
+        let target_mapping = env::var("RUST_TEST_ACTIVE_MAPPING").unwrap();
+        let active_mapping = mappings.get(&target_mapping).unwrap();
+
+        Self {
+            cmd: cmd.to_string(),
+            args: vec![],
+            mappings: active_mapping.clone(),
+        }
+    }
+
+    /// Lookup command in the mapping table
+    fn get_response(&self) -> (String, i32) {
+        let full_command = format!("{} {}", self.cmd, self.args.join(" "));
+        let entry = self
+            .mappings
+            .iter()
+            .find(|entry| entry.command == full_command)
+            .unwrap();
+        (entry.response.clone(), entry.code)
+    }
+}
+
+impl CommandLike for MockCommand {
+    fn arg(&mut self, arg: &str) -> &mut dyn CommandLike {
+        self.args.push(arg.to_string());
+        self
+    }
+
+    fn output(&mut self) -> io::Result<Output> {
+        let response = self.get_response();
+        println!(
+            "[MockCommand] Simulated command: {} {}",
+            self.cmd,
+            self.args.join(" ")
+        );
+
+        Ok(Output {
+            status: std::process::ExitStatus::from_raw(response.1),
+            stdout: response.0.as_bytes().to_vec(),
+            stderr: vec![],
+        })
+    }
+}
+
+/// Mock factory used to run argo/helm in tests
+pub struct MockCommandFactory;
+
+/// New command trait fo the mock factory
+impl CommandFactory for MockCommandFactory {
+    fn new_command(&self, cmd: &str) -> Box<dyn CommandLike> {
+        Box::new(MockCommand::new(cmd))
+    }
+}
+
+/// Get an instance of the CLI builder
+pub fn get_command_factory() -> Box<dyn CommandFactory> {
+    match env::var("MOCK_COMMAND") {
```

Comment:
It would be best to "namespace" any special environment variables that our application depends on.

eg WORKFLOWS_CLI_TEST_ENABLE_MOCK_COMMAND or similar

Reply:
Why did you choose to have the mocking live in the real application and not exist only in the test module?

Is the intention to have the ability to run the tests using the real "argo" and "helm" commands if they are installed on the system?